File: haskell/glue/src/Glue/Lib/Math/Arithmetic/Add.hs
module Glue.Lib.Math.Arithmetic.Add where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

-- Addition function
-- Mirrors Haskell Glue.Lib.Math.Arithmetic.Add.add exactly
add :: IR Eval
add = NativeFunc addImpl

-- Addition function implementation
-- Mirrors Haskell Glue.Lib.Math.Arithmetic.Add.addImpl exactly
addImpl :: IR Eval -> Eval (IR Eval)
addImpl left = pure $ NativeFunc (addTo left)

addTo :: IR Eval -> IR Eval -> Eval (IR Eval)
addTo left right = case (left, right) of
    (Integer a, Integer b) -> pure $ Integer (a + b)
    (Integer a, Float b) -> pure $ Float (fromIntegral a + b)
    (Float a, Integer b) -> pure $ Float (a + fromIntegral b)
    (Float a, Float b) -> pure $ Float (a + b)
    _ -> throwError $ wrongArgumentType ["number"]

File: dart/glue/lib/src/lib/math/arithmetic/add.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Addition function
/// Mirrors Haskell Glue.Lib.Math.Arithmetic.Add.add exactly
final Ir add = IrNativeFunc(addImpl);

/// Addition function implementation
/// Mirrors Haskell Glue.Lib.Math.Arithmetic.Add.addImpl exactly
Eval<Ir> addImpl(Ir left) {
  return Eval.pure(IrNativeFunc(addTo(left)));
}

/// Helper function for second argument
/// Mirrors Haskell Glue.Lib.Math.Arithmetic.Add.addTo exactly
Eval<Ir> Function(Ir) addTo(Ir left) {
  return (Ir right) => switch ((left, right)) {
    (IrInteger(value: final a), IrInteger(value: final b)) => Eval.pure(
      IrInteger(a + b),
    ),
    (IrInteger(value: final a), IrFloat(value: final b)) => Eval.pure(
      IrFloat(a + b),
    ),
    (IrFloat(value: final a), IrInteger(value: final b)) => Eval.pure(
      IrFloat(a + b),
    ),
    (IrFloat(value: final a), IrFloat(value: final b)) => Eval.pure(
      IrFloat(a + b),
    ),
    _ => throwError(wrongArgumentType(['number'])),
  };
}
